File: Backend/Services/parent/src/main/java/com/example/parent/service/TicketService.java
package com.example.parent.service;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import com.example.parent.entity.Ticket;
import com.example.parent.repository.TicketRepository;

@Service
public class TicketService {
    
    @Autowired
    private TicketRepository ticketRepository;
    
    @Autowired
    private RestTemplate restTemplate;
    
    private static final String ADMIN_SERVICE_URL = "http://localhost:8090";
    
    public List<Ticket> getTicketsByParentId(Long parentId) {
        return ticketRepository.findByParentIdOrderByCreatedAtDesc(parentId);
    }
    
    public Optional<Ticket> getTicketById(String ticketId) {
        return ticketRepository.findById(ticketId);
    }
    
    public Ticket createTicket(Long parentId, String subject, String description, String priority) {
        Ticket ticket = new Ticket();
        ticket.setId(UUID.randomUUID().toString());
        ticket.setParentId(parentId);
        ticket.setSubject(subject);
        ticket.setDescription(description);
        ticket.setPriority(priority);
        ticket.setStatus("OPEN");
        ticket.setCreatedAt(LocalDateTime.now());
        ticket.setUpdatedAt(LocalDateTime.now());
        ticket.setMessages(new ArrayList<>());
        
        // Assign random admin
        Long adminId = getRandomAdminId();
        ticket.setAdminId(adminId);
        
        return ticketRepository.save(ticket);
    }
    
    public Ticket addMessage(String ticketId, Long senderId, String senderType, String content) {
        Optional<Ticket> ticketOpt = ticketRepository.findById(ticketId);
        if (ticketOpt.isPresent()) {
            Ticket ticket = ticketOpt.get();
            
            Ticket.Message message = new Ticket.Message();
            message.setId(UUID.randomUUID().toString());
            message.setSenderId(senderId);
            message.setSenderType(senderType);
            message.setContent(content);
            message.setTimestamp(LocalDateTime.now());
            message.setRead(false);
            
            if (ticket.getMessages() == null) {
                ticket.setMessages(new ArrayList<>());
            }
            ticket.getMessages().add(message);
            ticket.setUpdatedAt(LocalDateTime.now());
            
            return ticketRepository.save(ticket);
        }
        return null;
    }
    
    public Ticket updateTicketStatus(String ticketId, String status) {
        Optional<Ticket> ticketOpt = ticketRepository.findById(ticketId);
        if (ticketOpt.isPresent()) {
            Ticket ticket = ticketOpt.get();
            ticket.setStatus(status);
            ticket.setUpdatedAt(LocalDateTime.now());
            
            if ("RESOLVED".equals(status) || "CLOSED".equals(status)) {
                ticket.setResolvedAt(LocalDateTime.now());
            }
            
            return ticketRepository.save(ticket);
        }
        return null;
    }
    
    public List<Ticket> getTicketsByStatus(String status) {
        return ticketRepository.findByStatusOrderByCreatedAtDesc(status);
    }
    
    public List<Ticket> getTicketsByAdminId(Long adminId) {
        return ticketRepository.findByAdminIdOrderByCreatedAtDesc(adminId);
    }
    
    private Long getRandomAdminId() {
        try {
            // Get all admin users from admin service
            String url = ADMIN_SERVICE_URL + "/api/admin/auth/all";
            AdminUser[] admins = restTemplate.getForObject(url, AdminUser[].class);
            
            if (admins != null && admins.length > 0) {
                // Select random admin
                int randomIndex = (int) (Math.random() * admins.length);
                return admins[randomIndex].getId();
            }
        } catch (Exception e) {
            System.err.println("Error fetching admins: " + e.getMessage());
        }
        
        // Fallback: return null if no admins found
        return null;
    }
    
    // Inner class for admin user data
    public static class AdminUser {
        private Long id;
        private String username;
        private String email;
        private String role;
        
        // Getters and setters
        public Long getId() { return id; }
        public void setId(Long id) { this.id = id; }
        public String getUsername() { return username; }
        public void setUsername(String username) { this.username = username; }
        public String getEmail() { return email; }
        public void setEmail(String email) { this.email = email; }
        public String getRole() { return role; }
        public void setRole(String role) { this.role = role; }
    }
}

File: Frontend/main-frontend/src/shared/services/ticket/ticketService.ts
// Ticket service for parent frontend
const PARENT_SERVICE_URL = "http://188.166.197.135:8082";

export interface Ticket {
  id: string;
  parentId: number;
  adminId?: number;
  subject: string;
  description: string;
  status: "OPEN" | "IN_PROGRESS" | "RESOLVED" | "CLOSED";
  priority: "LOW" | "MEDIUM" | "HIGH" | "URGENT";
  createdAt: string;
  updatedAt: string;
  resolvedAt?: string;
  messages: Message[];
}

export interface Message {
  id: string;
  senderId: number;
  senderType: "PARENT" | "ADMIN";
  content: string;
  timestamp: string;
  isRead: boolean;
}

export interface CreateTicketRequest {
  parentId: number;
  subject: string;
  description: string;
  priority: string;
}

export const ticketService = {
  // Get all tickets for a parent
  async getTicketsByParentId(parentId: number): Promise<Ticket[]> {
    try {
      const response = await fetch(
        `${PARENT_SERVICE_URL}/api/tickets/parent/${parentId}`,
        {
          credentials: "include",
        }
      );

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching tickets:", error);
      return [];
    }
  },

  // Get a specific ticket by ID
  async getTicketById(ticketId: string): Promise<Ticket | null> {
    try {
      const response = await fetch(
        `${PARENT_SERVICE_URL}/api/tickets/${ticketId}`,
        {
          credentials: "include",
        }
      );

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching ticket:", error);
      return null;
    }
  },

  // Create a new ticket
  async createTicket(request: CreateTicketRequest): Promise<Ticket | null> {
    try {
      const response = await fetch(`${PARENT_SERVICE_URL}/api/tickets`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        credentials: "include",
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error creating ticket:", error);
      return null;
    }
  },

  // Add a message to a ticket
  async addMessage(
    ticketId: string,
    content: string,
    senderId: number
  ): Promise<Ticket | null> {
    try {
      const response = await fetch(
        `${PARENT_SERVICE_URL}/api/tickets/${ticketId}/messages`,
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          credentials: "include",
          body: JSON.stringify({
            senderId,
            senderType: "PARENT",
            content,
          }),
        }
      );

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error adding message:", error);
      return null;
    }
  },
};
